services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -0,0 +1,388 @@
+        if (nonce == null) {
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_REQUEST));
+        }
+
+        AuthenticatedClientSessionModel clientSession = getAuthenticatedClientSession();
+
+        String note = clientSession.getNote(nonce);
+        if (note == null) {
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_REQUEST));
+        }
+
+        SupportedCredentialConfiguration offeredCredential;
+        try {
+            offeredCredential = objectMapper.readValue(note,
+                    SupportedCredentialConfiguration.class);
+            LOGGER.debugf("Creating an offer for %s - %s", offeredCredential.getScope(),
+                    offeredCredential.getFormat());
+            clientSession.removeNote(nonce);
+        } catch (JsonProcessingException e) {
+            LOGGER.errorf("Could not convert SupportedCredential JSON to POJO: %s", e);
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_REQUEST));
+        }
+
+        String preAuthorizedCode = generateAuthorizationCodeForClientSession(clientSession);
+
+        CredentialsOffer theOffer = new CredentialsOffer()
+                .setCredentialIssuer(OID4VCIssuerWellKnownProvider.getIssuer(session.getContext()))
+                .setCredentialConfigurationIds(List.of(offeredCredential.getId()))
+                .setGrants(
+                        new PreAuthorizedGrant()
+                                .setPreAuthorizedCode(
+                                        new PreAuthorizedCode()
+                                                .setPreAuthorizedCode(preAuthorizedCode)));
+
+        LOGGER.debugf("Responding with offer: %s", theOffer);
+        return Response.ok()
+                .entity(theOffer)
+                .build();
+    }
+
+    /**
+     * Returns a verifiable credential
+     */
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path(CREDENTIAL_PATH)
+    public Response requestCredential(
+            CredentialRequest credentialRequestVO) {
+        LOGGER.debugf("Received credentials request %s.", credentialRequestVO);
+
+        // do first to fail fast on auth
+        UserSessionModel userSessionModel = getUserSessionModel();
+
+        Format requestedFormat = credentialRequestVO.getFormat();
+        String requestedCredential = credentialRequestVO.getCredentialIdentifier();
+
+        SupportedCredentialConfiguration supportedCredentialConfiguration = Optional
+                .ofNullable(OID4VCIssuerWellKnownProvider.getSupportedCredentials(this.session)
+                        .get(requestedCredential))
+                .orElseThrow(
+                        () -> {
+                            LOGGER.debugf("Unsupported credential %s was requested.", requestedCredential);
+                            throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));
```

Comment:
`throw new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));`

might be

`return new BadRequestException(getErrorResponse(ErrorType.UNSUPPORTED_CREDENTIAL_TYPE));`

Reply:
Fixed in [0d88edc](https://github.com/keycloak/keycloak/pull/27931/commits/0d88edc5952e78aed03bc8acaaaa3f4674de7b5a)

Reply:
Thank you.